File: core/tools/xmp/registry.py
"""Namespace prefix ↔ URI registry shared across the XMP layer.

Prefixes registered here are recognized by :mod:`core.tools.xmp.packet` when
parsing existing XMP and emitted as ``xmlns:<prefix>="<uri>"`` declarations
when serializing. Unknown prefixes encountered while parsing existing files
are preserved verbatim (their URIs are stored on the parsed :class:`XmpData`
instance) so round-trips do not drop third-party namespaces.
"""

from __future__ import annotations

RDF_NS = "http://www.w3.org/1999/02/22-rdf-syntax-ns#"
DC_NS = "http://purl.org/dc/elements/1.1/"
X_NS = "adobe:ns:meta/"

_REGISTRY: dict[str, str] = {
    "rdf": RDF_NS,
    "dc": DC_NS,
}


def register_namespace(prefix: str, uri: str) -> None:
    """Register a prefix → URI mapping for parse/serialize."""
    if not prefix or not uri:
        raise ValueError("prefix and uri must be non-empty")
    _REGISTRY[prefix] = uri


def uri_for(prefix: str) -> str | None:
    return _REGISTRY.get(prefix)


def prefix_for(uri: str) -> str | None:
    for p, u in _REGISTRY.items():
        if u == uri:
            return p
    return None


def all_namespaces() -> dict[str, str]:
    return dict(_REGISTRY)


# Built-in tagdb namespaces. WD-Tagger has been writing the wdtag namespace
# since v4.x; sweep is new in v4.139.x and is owned by Sweep runs across the
# image-generation bridges.
register_namespace("wdtag", "http://ns.yu-ai-manager/wdtag/1.0/")
register_namespace("sweep", "http://ns.yu-ai-manager/sweep/1.0/")

File: crates/xmp-core/src/registry.rs
//! Namespace URI registry (mirrors Python core/tools/xmp/registry.py).
//! `wdtag` and `sweep` are the two namespaces this application writes.

const REGISTRY: &[(&str, &str)] = &[
    ("wdtag", "http://ns.yu-ai-manager/wdtag/1.0/"),
    ("sweep", "http://ns.yu-ai-manager/sweep/1.0/"),
    ("dc", "http://purl.org/dc/elements/1.1/"),
];

pub const RDF_NS: &str = "http://www.w3.org/1999/02/22-rdf-syntax-ns#";

/// Look up the URI for a known prefix (wdtag/sweep/dc). Returns `None` for
/// unregistered prefixes — callers fall back to synthetic `ns{n}` prefixes
/// (see `packet::parse`'s unknown-namespace handling).
pub fn uri_for(prefix: &str) -> Option<&'static str> {
    REGISTRY.iter().find(|(p, _)| *p == prefix).map(|(_, u)| *u)
}

/// Reverse lookup: URI to its registered prefix.
pub fn prefix_for(uri: &str) -> Option<&'static str> {
    REGISTRY.iter().find(|(_, u)| *u == uri).map(|(p, _)| *p)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn uri_for_returns_registered_namespace() {
        assert_eq!(uri_for("wdtag"), Some("http://ns.yu-ai-manager/wdtag/1.0/"));
        assert_eq!(uri_for("sweep"), Some("http://ns.yu-ai-manager/sweep/1.0/"));
    }

    #[test]
    fn uri_for_returns_none_for_unknown_prefix() {
        assert_eq!(uri_for("nope"), None);
    }

    #[test]
    fn prefix_for_is_the_reverse_of_uri_for() {
        assert_eq!(
            prefix_for("http://ns.yu-ai-manager/wdtag/1.0/"),
            Some("wdtag")
        );
        assert_eq!(prefix_for("http://example.com/unknown/"), None);
    }
}
